Add library doc to guides (#27581)
angular/angular: aio/content/guide/libraries-overview.md

# Using and Creating Libraries

Many applications need to solve the same general problems, such as presenting a unified user interface, presenting data, and allowing data entry.
Developers can create general solutions for particular domains that can be adapted for re-use in different apps. Such a solution can be published and shared as a *library*.

> Libraries extend Angular's base functionality. The forms and animations subsystems, for example, are libraries.

Reviewer: Animations got lots of special handling in core and compiler, which is not a real (user-implementable) library, I'd suggest use a different package as example.

Developers can create general solutions for particular domains that can be adapted for re-use in different apps. Such a solution can be published and shared as a *library*.

An Angular library is an Angular [project](guide/glossary#project) that differs from an app in that it cannot run on its own.
A library must be imported and used in an app.

Libraries extend Angular's base functionality. For example, to add [reactive forms](guide/reactive-forms) to an app. you import the `ReactiveFormsModule` from the library `@angular/forms`.
Adding the [service worker](guide/service-worker-intro) library to an Angular application is one of the steps for turning an application into a [Progressive Web App](https://developers.google.com/web/progressive-web-apps/) (PWA).
[Angular Material](https://material.angular.io/) is an example of a large, general-purpose library that provides sophisticated, reusable, and adaptable UI components.

Any app developer can use these and other libraries that have been published as npm packages by the Angular team or by third parties. See [Using Published Libraries](guide/use-libs).

## Creating libraries

> If you have developed functionality that is suitable for reuse, you can create your own libraries. You can create libraries for your own use, and also publish them as [npm packages](guide/npm-packages) to share them with other Angular developers. See [Creating Libraries](guide/create-libs).

Reviewer 2: Many Angular developers are on Enterprise teams. Many of these team's run their own npm Enterprise on-premise or they run a package management system, like Sonatype Nexus or JFrog Artifactory, on-premise or in their own data center. I think that we should make it clear that we support all of these configurations.

I think that this covers it:
"If you have developed functionality that is suitable for reuse, you can create your own libraries. These libraries can be used locally in your workspace. Optionally, you can also organize them as [npm packages](guide/npm-packages) to share with other Angular developers. These packages can be published to the npm registry, a private npm Enterprise registry, or a private package management system that supports npm packages."

> # Overview of Angular Libraries

Reviewer 3: I would prefer if this file was called "libraries.md" so that we have a nice link https://angular.io/guide/libraries. can you please rename? the h1 title above can remain as is.

Many applications need to solve the same general problems, such as presenting a unified user interface, presenting data, and allowing data entry.

> Developers can create general solutions for particular domains that can be adapted for re-use in different apps. Such a solution can be published and shared as a *library*.

Reviewer 3: ```suggestion
Developers can create general solutions for particular domains that can be adapted for re-use in different apps. Such a solution can be built as Angular *libraries* and these libraries can be published and shared as *npm packages*.
```

An Angular library is an Angular [project](guide/glossary#project) that differs from an app in that it cannot run on its own.
A library must be imported and used in an app.

> Libraries extend Angular's base functionality. For example, to add [reactive forms](guide/reactive-forms) to an app. you import the `ReactiveFormsModule` from the library `@angular/forms`.

Reviewer 3: ```suggestion
Libraries extend Angular's base functionality. For example, to add [reactive forms](guide/reactive-forms) to an app, you `ng add @angular/forms` package and import the `ReactiveFormsModule` from the library `@angular/forms` in your application code.
```

> Adding the [service worker](guide/service-worker-intro) library to an Angular application is one of the steps for turning an application into a [Progressive Web App](https://developers.google.com/web/progressive-web-apps/) (PWA).

Reviewer 3: ```suggestion
Similarly, by adding the [service worker](guide/service-worker-intro) library to an Angular application is one of the steps for turning an application into a [Progressive Web App](https://developers.google.com/web/progressive-web-apps/) (PWA).
```

[Angular Material](https://material.angular.io/) is an example of a large, general-purpose library that provides sophisticated, reusable, and adaptable UI components.

Any app developer can use these and other libraries that have been published as npm packages by the Angular team or by third parties. See [Using Published Libraries](guide/use-libs).

## Creating libraries

If you have developed functionality that is suitable for reuse, you can create your own libraries.

> These libraries can be used locally in your workspace.

Reviewer 3: ```suggestion
These libraries can be used locally in your workspace, or you can publish them as [npm packages](guide/npm-packages) to share with other projects or other Angular developers.
```

> Optionally, you can also organize them as [npm packages](guide/npm-packages)to share with other Angular developers.

Reviewer 3: ```suggestion
```

These packages can be published to the npm registry, a private npm Enterprise registry, or a private package management system that supports npm packages. 
See [Creating Libraries](guide/create-libs).

Whether you decide to package functionality as a library is an architectural decision, similar to deciding whether a piece of functionality is a component or a service, or deciding on the scope of a component.

Packaging functionality as a library forces the artifacts in the library to be decoupled from the application's business logic.
This can help to avoid various bad practices or architecture mistakes that can make it difficult to decouple and reuse code in the future.

Putting code into a separate library, however, is more complex than simply putting everything in one app.
It requires more of an investment in time and thought for managing, maintaining, and updating the library.
This complexity only pays off when the library is being used in multiple apps.

> You might not want to waste resources putting code into a library just because it might one day see some reuse.

Reviewer 3: this sounds too negative. consider rewording...
